### PR TITLE
Add Software ID in opam file

### PR DESCRIPTION
**Pull request by rjbou**
Add Software Heritage ID  field `swhid` in url section, and in opam show, and fallback support

### CODE REVIEW

File: src/core/opamHash.ml [hunk 1]
```diff
@@ -148,3 +148,50 @@ end
...
+  let of_json = function
+    | `String s -> of_string_opt s
+    | _ -> None
+
+  (** XXX Keep that? *)
+  let compare = String.compare
+  let equal = String.equal
+
+  module O = struct
+    type _t = t
```
Comment (zapashcanon) [inline src/core/opamHash.ml:186]: `type nonrec t = t` ?

File: src/core/opamUrl.ml [hunk 2]
```diff
@@ -308,3 +308,21 @@ module Op = struct
...
+module SWHID = struct
+
+  let prefix = "swhid.opam.ocaml.org/"
+  let is_valid u =
+    u.backend = `http
+    && (String.equal u.transport "http" || String.equal u.transport "https")
+    && OpamStd.String.starts_with ~prefix u.path
+  let of_url u =
+    try
+      Some (OpamHash.SWHID.of_string (OpamStd.String.remove_prefix ~prefix u.path))
```
Comment (zapashcanon) [inline src/core/opamUrl.ml:321]: IIUC this doesn't support having directly `swh:...` as an URL ? I think we should support both *now* (even if in the beginning we use only `http{s}://swhid.opam.ocaml.org/...` in the official opam repository) to ease the transition to `swh:...` only - which is registered at IANA etc.

Comment (rjbou) [reply src/core/opamUrl.ml]: Tried with simply `swh:...`, it is resolved to a ssh url. There will be another url kind to handle, and besides that, older opams will print the ssh url.

File: src/core/opamSWHID.ml [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  try
+    Some (of_string (OpamStd.String.remove_prefix ~prefix url.OpamUrl.path))
+  with Invalid_argument _ -> None
+
+let to_url swh =
+  let path = Printf.sprintf "%s%s" prefix (to_string swh) in
+  OpamUrl.{ transport = "https"; backend = `http ; hash = None; path }
+
+
+(** SWHID retrieval functions *)
```
Comment (zapashcanon) [inline src/core/opamSWHID.ml]: You can remove everything in this file starting from this line. :)

**Comment by rjbou** [comment]
Updated!

File: tests/reftests/swhid.unix.test [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  checksum: "md5=039a77d9ffb9b22ab845df30436390c1"
+}
+### opam switch create fallback --empty
+### opam install snappy-ok
+The following actions will be performed:
+=== install 1 package
+  - install snappy-ok 1
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> retrieved snappy-ok.1  (https://ygrek.org/p/release/ocaml-snappy/ocaml-snappy-0.1.2.tar.gz)
```
[the remark this answers is not in the document]
Comment (rjbou) [reply tests/reftests/swhid.unix.test:20]: No other solution, i'm afraid, we need to retrieve the archive from swh. The tests don't pass the CI in any way, as we depend on distant opam repo in some other tests, and then need network, it is not a new issue.

File: src/client/opamListCommand.ml [hunk 5]
```diff
@@ -540,8 +540,24 @@ let detail_printer ?prettify ?normalise ?(sort=false) st nv =
...
+           | None -> raise Not_found
+         else
+         let opam =
+           if f = "url.mirrors" then
+             OpamFile.(OPAM.with_url_opt
+                         (OPAM.url opam >>| URL.with_swhid_opt None)
+                         opam)
+           else opam
+         in
+         List.assoc f (OpamFile.OPAM.to_list opam)
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/client/opamListCommand.ml:558]: Agree, factorisation habit, but sometimes in less clear :)
No better solution than proposed

**Review by kit-ty-kate — APPROVED** [review]

File: src/repository/opamDownload.ml [hunk 6]
```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
...
+      "obj_id": "4453cfbdab1a996658cd1a815711664ee7742380"
+     }
+  *)
+
+  let get_output ?(post=false) url =
+    get_output ~post url @@| function
+    | Some out ->
+      Some (String.concat "" out)
+    | None ->
+      OpamConsole.error "Software Heritage fallback need %s or %s installed"
```
Comment (dra27) [inline src/repository/opamDownload.ml:259]:
```suggestion
      OpamConsole.error "Software Heritage fallback needs %s or %s installed"
```

File: src/repository/opamDownload.ml [hunk 7]
```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
...
+  let fallback_err fmt = Printf.sprintf ("SWH fallback: "^^fmt)
+
+  let get_url ?(timeout=6) swhid =
+    let attempts = timeout in
+    let hash = OpamSWHID.hash swhid in
+    let rec aux timeout =
+      if timeout <= 0 then
+        Done (Not_available
+                (Some (fallback_err "timeout"),
+                 fallback_err "%d attempts tried, aborting" attempts))
```
Comment (dra27) [inline src/repository/opamDownload.ml:288]:
```suggestion
                 fallback_err "%d attempts tried; aborting" attempts))
```

File: src/repository/opamDownload.ml [hunk 8]
```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
...
+  (* for the moment only used in sources, not extra sources or files *)
+  let archive_fallback ?timeout urlf dirnames =
+    match OpamFile.URL.swhid urlf with
+    | None -> Done (Result None)
+    | Some swhid ->
+      if check_post_tool () then
+        (* Add a global modifier and/or command for default answering *)
+        if OpamConsole.confirm ~default:false
+            "Source %s is not available. Do you want to try to retrieve it \
+             from Software Heritage cache? It may take few minutes."
```
Comment (dra27) [inline src/repository/opamDownload.ml:309]: Possibly include a link to https://www.softwareheritage.org for information about what it is?

File: src/repository/opamDownload.ml [hunk 9]
```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
...
+                     Result (Some "SWH fallback"))
+                  else
+                    Not_available (
+                      Some (fallback_err "archive not valid"),
+                      fallback_err
+                        "archive corrupted, opam file swhid %S vs computed %S"
+                        hash computed)))
+        else
+          Done (Not_available
+                  (Some (fallback_err "no retrieval"),
```
Comment (dra27) [inline src/repository/opamDownload.ml:354]: Possibly "skip" rather than "no"?

**Review by dra27 — COMMENTED** [review]
Some suggestions on messages

File: src/repository/opamDownload.ml [hunk 10]
```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
...
+let get_output ~post ?(args=[]) url =
+  let cmd_args =
+    download_args ~url ~out:"-" ~retry:OpamRepositoryConfig.(!r.retries)
+      ~compress:false ()
+    @ args
+  in
+  let cmd_args =
+    if post then
+      match cmd_args with
+      | ("wget"|"curl" as cmd)::args -> Some (cmd :: ["-X"; "POST"] @ args)
```
Comment (AltGr) [inline src/repository/opamDownload.ml:215]: It's probably time we do the http requests through an OCaml library!
(unrelated to the nice work here, I know :) )

File: src/repository/opamDownload.ml [hunk 11]
```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
...
+      Some (match status with
+          | "done" -> `Done (OpamUrl.of_string fetch_url)
+          | "pending" -> `Pending
+          | "new" -> `New
+          | "failed" -> `Failed
+          | _ -> `Unknown)
+
+  let fallback_err fmt = Printf.sprintf ("SWH fallback: "^^fmt)
+
+  let get_url ?(timeout=6) swhid =
```
Comment (AltGr) [inline src/repository/opamDownload.ml:281]: Nitpicking I know, but this is not technically a timeout but a `max_tries`

**Review by AltGr — CHANGES_REQUESTED** [review]
Very nice work!

I have one concern: IIUC, the fallback download handling is triggered as soon as the initial download fails. Isn't this a problem for scenarios where the network is down ?
A solution might be to ping `softwareheritage.org` first ?

File: src/repository/opamDownload.ml (hunk 10 above)
Comment (rjbou) [reply src/repository/opamDownload.ml:215]: agree! It's quite hackish, working but hackish

**Comment by kit-ty-kate** [comment]
> A solution might be to ping `softwareheritage.org` first ?

a real ping isn’t a good solution as some (weird but real) networks disallow ICMP requests

File: src/repository/opamDownload.ml (hunk 10 above)
Comment (rjbou) [reply src/repository/opamDownload.ml:215]: yes, this is in the todolist with other libs (tar/git/etc.)

**Comment by rjbou** [comment]
In fact, we can simply retrieve the default page and drop the result.

File: src/repository/opamDownload.ml [hunk 12]
```diff
@@ -225,6 +225,7 @@ module SWHID = struct
 
   let log fmt = OpamConsole.log "CURL(SWHID)" fmt
 
+  let main_url = OpamUrl.of_string "http://softwareheritage.com"
```
Comment (kit-ty-kate) [inline src/repository/opamDownload.ml:228]:
```
This domain name has been registered with Gandi.net
```

Comment (rjbou) [reply src/repository/opamDownload.ml:228]: wrong url indedd

Comment (kit-ty-kate) [reply src/repository/opamDownload.ml]: what about just sending a request such as `curl -X POST https://archive.softwareheritage.org/api/1`. It returns no output nor failure?

Comment (rjbou) [reply src/repository/opamDownload.ml]: In that way, we can put the check directly on fallback mechanism, not in `OpamUpdate`. :+1:

Comment (kit-ty-kate) [reply src/repository/opamDownload.ml]: even better, they have an api endpoint just for that: https://docs.softwareheritage.org/devel/apidoc/swh.web.api.views.ping.html?highlight=ping#get--api-1-ping-

**Review by kit-ty-kate — APPROVED** [review]

File: tests/reftests/swhid.unix.test [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+### opam install snappy-swhid-ko
+The following actions will be performed:
+=== install 1 package
+  - install snappy-swhid-ko 2
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+Source https://fa.ke/url.tar.gz is not available. Do you want to try to retrieve it from Software Heritage cache (https://www.softwareheritage.org)? It may take few minutes. [y/n] y
+[ERROR] Failed to get sources of snappy-swhid-ko.2: SWH fallback: Unknown swhid
+
+OpamSolution.Fetch_fail("SWH fallback: Unknown swhid")
```
Comment (AltGr) [inline tests/reftests/swhid.unix.test:111]: Extra exn print ?

**Review by AltGr — APPROVED** [review]

File: tests/reftests/swhid.unix.test (hunk 13 above)
Comment (rjbou) [reply tests/reftests/swhid.unix.test:111]: Already existing issue, see l.39

**Comment by avsm** [comment]
For my own education, is there any documentation about:
- what Software Heritage IDs are
- how they are added to an opam package
- why they are different from a conventional content-addressed store (e.g. something like IPFS/hash fetching)
- can the fallback be switched off, and/or does this affect purely offline use of opam with local files?

Seems like a nice feature, but I'm missing a little background.

**Comment by dra27** [comment]
@avsm - apologies, this got discussed at a dev meeting in September, but only put into notes from the meeting, and not commented back (which is my fault...)!

> what Software Heritage IDs are
The SWHID (see also [their documentation](https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html)) is just a(nother) content hash.

> how they are added to an opam package

This PR (and opam 2.2) is focussed on being able to consume the content hashes - for adding them, the intention is to have existing package submission tools (`opam-publish`, `dune-release`, etc.) generate the SWHID along with other hashes from the release tarball. opam-repository _may_ then choose to lint based on these. Part of the Software Heritage project work for programming language ecosystems is that they monitor opam-repository and ensure that the archives referenced are actually in their archive (it is also possible - and desirable - that the package submission would itself submit the archive to SWH, but the point is that the SWHID being in an opam file is neither gated nor requires its actually being submitted to the service).

Adding the facility to use the service in opam 2.2 obviously doesn't mandate deploying it for all packages on opam-repository (or in a future version) - just as adding the additional sha hash functions earlier in opam 1.x didn't.

> why they are different from a conventional content-addressed store (e.g. something like IPFS/hash fetching)

They're related, indeed largely equivalent. The Software Heritage project, though, is focussed on persistence, where IPFS gives availability. I don't have an answer as to why they end up using different content hashes - although there is a [project in progress](https://www.softwareheritage.org/2022/02/10/building-bridge-to-the-software-heritage-archive/) to provide the Software Heritage archive over IPFS.

> can the fallback be switched off, and/or does this affect purely offline use of opam with local files?

Yes - in fact, unless I've misread, it is not automatically enabled (it's an interactive prompt and would be off in batch/non-interactive mode, @rjbou?).

opam being able to retrieve archives from Software Heritage is a partial solution to an upstream server being down (opam.ocaml.org, GitHub, etc.) but that's not the primary point. Primarily, this provides an external solution and ongoing solution for things like https://github.com/ocaml/opam-source-archives and should mean that we never end up scrabbling to find archives in old server images etc. when gforge services get shut down or personal web servers get taken offline, etc.

**Comment by rjbou** [comment]
> can the fallback be switched off, and/or does this affect purely offline use of opam with local files?

The fallback is enabled by default, it can be disabled by `opam option swh_fallback=false`. It is used in last resort when archive is unreachable via 1. url defined in opam package 2. in opam repository cache & its mirrors. From there, a prompt asks if you want to try download via SWH (see [in test](https://github.com/rjbou/opam/blob/swhid2/tests/reftests/swhid.unix.test#L88-L90)). Before calling the fallback/prompting, we ensure that network is up & SWH api is up.

**Comment by avsm** [comment]
Many thanks for the updates. One quick design question though, after reading the SWHID docs linked: it looks like they [define a URL schema](https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html#uri-scheme), so why don't we just interpret this schema in `url` fields (just as we do for https+git for example)?

**Comment by zapashcanon** [comment]
We wanted to use it but we couldn't, mainly for compatibility with old clients. See [this comment](https://github.com/ocaml/opam/pull/4859#discussion_r778476033).

**Comment by rdicosmo** [comment]
Happy to see this moving forward!
As a side note, a working group is assembling to start the normalization process of SWHIDs, and it would be great if some of the contributors to this work in opam could join. See more information here https://www.swhid.org/

**Review by dra27 — APPROVED** [review]

**Pull request merged**